modules/map/tools/map_datachecker/server/loops_verify_agent.cc

```diff
@@ -0,0 +1,299 @@
+    return;
+  }
+
+  if (sp_laps_checker_->get_return_state() != 0) {
+    response->set_progress(0.0);
+    response->set_code(sp_laps_checker_->get_return_state());
+    return;
+  }
+
+  double progress = sp_laps_checker_->get_progress();
+  response->set_progress(progress);
+  response->set_code(ErrorCode::SUCCESS);
+  if (std::abs(1.0 - progress) < 1e-8) {
+    double confidence = sp_laps_checker_->get_confidence();
+    size_t lap = sp_laps_checker_->get_lap();
+    AINFO << "aquired lap: " << lap << ", conf: " << confidence;
+    apollo::hdmap::LoopResult *loop_result = response->mutable_loop_result();
+
+    loop_result->set_loop_num(static_cast<double>(lap));
+    bool is_reached = lap >= get_loops_to_check(request) ? true : false;
+    loop_result->set_is_reached(is_reached);
+
+    DataType data_type = request->type();
+    if (data_type == DataType::MAP_CHECKOUT) {
+      if (is_reached) {
+        loop_result->set_loop_num(
+            static_cast<double>(sp_conf_->laps_number_additional));
+      } else {
+        loop_result->set_loop_num(
+            lap - sp_conf_->laps_number >= 0
+                ? static_cast<double>(lap - sp_conf_->laps_number)
+                : 0.0);
+      }
+    }
+  }
+}
+
+void LoopsVerifyAgent::StopVerify(LoopsVerifyRequest *request,
+                                  LoopsVerifyResponse *response) {
+  AINFO << "call StopVerify";
+  response->set_code(ErrorCode::SUCCESS);
+  set_state(LoopsVerifyAgentState::IDLE);
+  if (sp_laps_checker_ == nullptr) {
+    response->set_progress(0.0);
+    return;
+  }
+  response->set_progress(sp_laps_checker_->get_progress());
+  if (std::abs(1.0 - sp_laps_checker_->get_progress()) < 1e-8) {
+    double conf = sp_laps_checker_->get_confidence();
+    size_t lap = sp_laps_checker_->get_lap();
+    AINFO << "aquired lap: " << lap << ", conf: " << conf;
+    apollo::hdmap::LoopResult *loop_result = response->mutable_loop_result();
+    loop_result->set_loop_num(static_cast<double>(lap));
+    bool is_reached = lap >= get_loops_to_check(request) ? true : false;
+    loop_result->set_is_reached(is_reached);
+
+    DataType data_type = request->type();
+    if (data_type == DataType::MAP_CHECKOUT) {
+      if (is_reached) {
+        loop_result->set_loop_num(
+            static_cast<double>(sp_conf_->laps_number_additional));
+      } else {
+        loop_result->set_loop_num(
+            lap - sp_conf_->laps_number >= 0
+                ? static_cast<double>(lap - sp_conf_->laps_number)
+                : 0.0);
+      }
+    }
+  }
+}
+
+std::shared_ptr<std::vector<std::pair<double, double>>>
+LoopsVerifyAgent::get_verify_range(LoopsVerifyRequest *request) {
+  std::shared_ptr<std::vector<std::pair<double, double>>> sp_range(
+      new std::vector<std::pair<double, double>>());
+  int range_size = request->range_size();
+  for (int i = 0; i < range_size; ++i) {
+    const ::apollo::hdmap::VerifyRange &range = request->range(i);
```

Comment:
Try `for (const VerifyRange& range : request->range())`, which is shorter, more efficient and self declaration.